cpp/include/cugraph/prims/gather_edges.cuh

```diff
@@ -0,0 +1,495 @@
+      // partition offsets because it is a concatenation of all the offsets
+      // across all partitions
+      auto location = location_in_segment + vertex_count_offsets[partition_id];
+
+      return global_out_degrees[location];
+    });
+  return active_source_degrees;
+}
+
+/**
+ * @brief Calculate global degree information for all vertices represented by current gpu
+ *
+ * Calculate local degree and perform row wise exclusive scan over all gpus in column
+ * communicator.
+ *
+ * @tparam GraphViewType Type of the passed non-owning graph object.
+ * @param handle RAFT handle object to encapsulate resources (e.g. CUDA stream, communicator, and
+ * handles to various CUDA libraries) to run graph algorithms.
+ * @param graph_view Non-owning graph object.
+ * @return Tuple of two device vectors. The first one contains per source edge-count encountered
+ * by gpus in the column communicator before current gpu. The second device vector contains the
+ * global out degree for every source represented by current gpu
+ */
+template <typename GraphViewType>
+std::tuple<rmm::device_uvector<typename GraphViewType::edge_type>,
+           rmm::device_uvector<typename GraphViewType::edge_type>>
+get_global_degree_information(raft::handle_t const& handle, GraphViewType const& graph_view)
+{
+  using edge_t       = typename GraphViewType::edge_type;
+  auto local_degrees = detail::calculate_local_degrees(handle, graph_view);
+
+  auto& col_comm      = handle.get_subcomm(cugraph::partition_2d::key_naming_t().col_name());
+  auto const col_size = col_comm.get_size();
+  auto const col_rank = col_comm.get_rank();
+  auto& row_comm      = handle.get_subcomm(cugraph::partition_2d::key_naming_t().row_name());
+  auto const row_size = row_comm.get_size();
+
+  auto& comm           = handle.get_comms();
+  auto const comm_size = comm.get_size();
+  auto const comm_rank = comm.get_rank();
+
+  rmm::device_uvector<edge_t> temp_input(local_degrees.size(), handle.get_stream());
+  raft::update_device(
+    temp_input.data(), local_degrees.data(), local_degrees.size(), handle.get_stream());
+
+  rmm::device_uvector<edge_t> recv_data(local_degrees.size(), handle.get_stream());
+  if (col_rank == 0) {
+    thrust::fill(handle.get_thrust_policy(), recv_data.begin(), recv_data.end(), edge_t{0});
+  }
+  for (int i = 0; i < col_size - 1; ++i) {
+    if (col_rank == i) {
+      comm.device_send(
+        temp_input.begin(), temp_input.size(), comm_rank + row_size, handle.get_stream());
+    }
+    if (col_rank == i + 1) {
+      comm.device_recv(
+        recv_data.begin(), recv_data.size(), comm_rank - row_size, handle.get_stream());
+      thrust::transform(handle.get_thrust_policy(),
+                        temp_input.begin(),
+                        temp_input.end(),
+                        recv_data.begin(),
+                        temp_input.begin(),
+                        thrust::plus<edge_t>());
+    }
+    handle.get_comms().barrier();
+  }
```

Comment:
Just FYI for future optimization.

So, in this code, only two GPUs (out of col_comm_size GPUs) will be active in each iteration, only one GPU has computation. This will become a scalability bottleneck especially when col_comm_size is large.

What you may do instead is

There are col_comm_size number of blocks in local_degrees.

col_comm_rank = 0 starts with the block0
col_comm_rank = 1 starts with the block1
and so on....

Then, each GPU will send the data for one block and receive and compute the data for one block as well in each iteration. Every GPU will do a comparable amount of send/receive and computation.

You may add a FIXME statement for future reference.

Reply:
I am not sure I fully understand your approach now but I'll get there. Having a more performant vector exclusive sum will decide if we can avoid gathering all of the global out degrees and global out offsets.